what4/src/What4/Solver/Z3.hs

```diff
@@ -288,18 +302,54 @@ runZ3Horn sym log_data inv_fns horn_clauses = do
             _ -> Nothing)
           (SMT2.sessionWriter session)
           (Syntax.getValue [])
-        SMT2.parseFnValues sym (SMT2.sessionWriter session) inv_fns sexp)
+        SMT2.parseFnValues sym (SMT2.sessionWriter session) lia_inv_fns sexp)
       return
-      result
+      check_sat_result
+
+  if do_bv_to_lia_transform then
+    case get_value_result of
+      Sat lia_defined_fns -> do
+        defined_inv_fns <- MapF.fromList <$> mapM
+          (\(SomeSymFn fn) ->
+            if| Just (SomeSymFn lia_fn) <- Map.lookup (SomeSymFn fn) bv_to_lia_fn_subst
+              , Just (SymFnWrapper lia_defined_fn) <- MapF.lookup (SymFnWrapper lia_fn) lia_defined_fns -> do
+                some_defined_fn <- transformSymFnLIA2BV sym $ SomeSymFn lia_defined_fn
+                case some_defined_fn of
+                  SomeSymFn defined_fn
+                    | Just Refl <- testEquality (fnArgTypes fn) (fnArgTypes defined_fn)
+                    , Just Refl <- testEquality (fnReturnType fn) (fnReturnType defined_fn) ->
+                      return $ MapF.Pair (SymFnWrapper fn) (SymFnWrapper defined_fn)
+                  _ -> fail $ "runZ3Horn: unexpected function " ++ show fn
+              | otherwise -> fail $ "runZ3Horn: unexpected function " ++ show fn)
+          inv_fns
+
+        all_unsat <- and <$> mapM
+          (\clause -> do
+            defined_clause <- notPred sym =<< substituteSymFns sym defined_inv_fns clause
+            runZ3InOverride sym (log_data { logVerbosity = 2 }) [defined_clause] $ return . isUnsat)
+          horn_clauses
+
+        return $ if all_unsat then Sat defined_inv_fns else Unknown
+
+      _ -> return Unknown
```

Comment:
It might be worth factoring this large subexpression out into its own function, as there is a lot of distance from the `  if do_bv_to_lia_transform then ...` part to the `... else return get_value_result` part.